### PR TITLE
(API) Added Biome Maps [1.18 preparation]

### PR DESCRIPTION
**Pull request by TheBusyBiscuit**
## Description

This pull requests introduces Biome maps, they have been loosely inspired by our existing Material tags but serve a greater purpose.
In the future, these can be used to allow extensive customization for server owners.
But right now, they will simply make our lives easier.

The reason for this addition is the fact that 1.18 changed all biome names.
A framework like this will help us stay backwards-compatible, we can simply create seperate biome-maps for pre-1.18 and post-1.18.

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [x] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Comment by WalshyDev** [comment]
Pog this is awesome, great work

**Comment by TheBusyBiscuit** [comment]
@Slimefun/code-reviewers I have marked this as "ready-for-review".
Some documentation is probably still needed but the overall framework should be finished.
I even got version-compatibility test cases going :)

Please give this a thorough review, when we got this sorted, the next step would be to add 1.18 biome-maps for all these resources in an upcoming 1.18 PR, combine that with the other 1.18 changes, dough and MockBukkit and we should be ready to go.

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/biomes/BiomeDataConverter.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * @see BiomeMap
+ */
+@FunctionalInterface
+public interface BiomeDataConverter<T> {
+
+    /**
+     * Override this method and provide a way to convert a {@link JsonElement} into your
+     * desired data type.
+     * <p>
+     * For primitive data values, you you can also just use the following method references:
```
Comment (Sfiguz7) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/biomes/BiomeDataConverter.java:25]:
```suggestion
     * For primitive data values, you can also just use the following method references:
```

**Review by Sfiguz7 — CHANGES_REQUESTED** [review]
Quite sure thebusybiscuit_wonderland won't exist for a while unfortunately :( But hopefully it will before walshy_desert

**Review by Sfiguz7 — DISMISSED** [review]
I could not find anything other than the typo so LGTM

**Review by svr333 — COMMENTED** [review]
besides the minor comment remarks it looks good to me too

**Review by svr333 — APPROVED** [review]
Looks good to me

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/biomes/TestBiomeMapCompatibility.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    void testCompatibilities(String name, MinecraftVersion version) {
+        assertCompatibility(name, version);
+    }
+
+    /**
+     * We manually specify here which biome map works on which minecraft versions.
+     * 
+     * @return A {@link Stream} of {@link Arguments} for our unit test.
+     */
+    private static @Nonnull Stream<Arguments> biomeMaps() {
```
Comment (WalshyDev) [inline src/test/java/io/github/thebusybiscuit/slimefun4/utils/biomes/TestBiomeMapCompatibility.java:98]: This method is currently unused, what's the purpose here?

**Review by WalshyDev — CHANGES_REQUESTED** [review]

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/biomes/TestBiomeMapCompatibility.java (hunk 2 above)
Comment (TheBusyBiscuit) [reply src/test/java/io/github/thebusybiscuit/slimefun4/utils/biomes/TestBiomeMapCompatibility.java:98]: Look above. And look at the method.
It's a `Stream<Arguments>` method.
It's used for parameterized unit tests.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/resources/NetherIceResource.java [hunk 3]
```diff
@@ -23,7 +25,12 @@
     NetherIceResource() {
         super("nether_ice", "Nether Ice", SlimefunItems.NETHER_ICE, 6, true);
 
-        biomes = getBiomeMap(this, "/biome-maps/nether_ice.json");
+        // 1.16+ introduced Nether biomes
+        if (Slimefun.getMinecraftVersion().isAtLeast(MinecraftVersion.MINECRAFT_1_16)) {
+            biomes = getBiomeMap(this, "/biome-maps/nether_ice.1_16.json");
+        } else {
+            biomes = getBiomeMap(this, "/biome-maps/nether_ice.json");
```
[the remark this answers is not in the document]
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/resources/NetherIceResource.java]: That's literally the convention we are going for here...
Seems like you accidentally reviewed an older state.

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/biomes/TestBiomeMapCompatibility.java (hunk 2 above)
Comment (WalshyDev) [reply src/test/java/io/github/thebusybiscuit/slimefun4/utils/biomes/TestBiomeMapCompatibility.java:98]: Oh whoops, I didn't realize you could do MethodSource and specify a name... huh neat.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/resources/NetherIceResource.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/resources/NetherIceResource.java]: Oh fuck my life... I opened this PR 2 days ago and GitHub didn't think to update it

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/biomes/BiomeMapParser.java [hunk 4]
```diff
@@ -40,6 +42,24 @@
     private final BiomeDataConverter<T> valueConverter;
     private final Map<Biome, T> map = new EnumMap<>(Biome.class);
 
+    /**
+     * This flag specifies whether the parsing is "lenient" or not.
+     * A lenient parser will not throw a {@link BiomeMapException} if the {@link Biome}
+     * could not be found.
+     * The default value is false.
+     */
+    private boolean isLenient = false;
```
[the remark this answers is not in the document]
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/biomes/BiomeMapParser.java:51]: Actually, not.

Lenient is false by default in production code, so as to properly log an exception for unknown biomes, incase someone modified the data file or so.

But I wanted to have the theoretical option to just skip invalid biomes, incase an addon developer does not feel like having different biome maps for different game versions.
It's not used at all right now, but the option to have a more lenient parser does exist code-wise. If an addon developer should care for that.

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/biomes/TestBiomeMapCompatibility.java (hunk 2 above)
Comment (TheBusyBiscuit) [reply src/test/java/io/github/thebusybiscuit/slimefun4/utils/biomes/TestBiomeMapCompatibility.java:98]: Yup, it is very neat!
I actually use it relatively often: https://github.com/Slimefun/Slimefun4/search?q=MethodSource

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/PatternUtils.java [hunk 5]
```diff
@@ -20,7 +20,8 @@ private PatternUtils() {}
 
     public static final Pattern SLASH_SEPARATOR = Pattern.compile(" / ");
 
-    public static final Pattern MINECRAFT_MATERIAL = Pattern.compile("minecraft:[a-z_]+");
+    public static final Pattern MINECRAFT_NAMESPACEDKEY = Pattern.compile("minecraft:[a-z_]+");
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/PatternUtils.java:23]: Worth noting that the namespace regex is actually `[a-z._-]+`. there are a lot of namespaces in MCs code using things like dots. I'm not sure if any are in Spigot yet but we should probably keep this regex inline with the actual regex for future-proofing
```suggestion
    public static final Pattern MINECRAFT_NAMESPACEDKEY = Pattern.compile("minecraft:[a-z._-]+");
```

**Review by WalshyDev — APPROVED** [review]
Apologies for my useless last review haha. One small comment but this LGTM

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/PatternUtils.java (hunk 5 above)
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/PatternUtils.java:23]: It's actually even `[a-z0-9/._-]+`, according to https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/browse/src/main/java/org/bukkit/NamespacedKey.java#35
```suggestion
    public static final Pattern MINECRAFT_NAMESPACEDKEY = Pattern.compile("minecraft:[a-z0-9/._-]+");
```

Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/PatternUtils.java:23]: Interesting, can't think of any with a slash off the top of my head.

**Review by WalshyDev — DISMISSED** [review]
LGTM

**Review by svr333 — APPROVED** [review]

**Comment by TerslenK** [comment]
what is this for?

**Comment by svr333** [comment]
> The reason for this addition is the fact that 1.18 changed all biome names.
A framework like this will help us stay backwards-compatible, we can simply create seperate biome-maps for pre-1.18 and post-1.18.

**Comment by TerslenK** [comment]
> > The reason for this addition is the fact that 1.18 changed all biome names.
> > A framework like this will help us stay backwards-compatible, we can simply create seperate biome-maps for pre-1.18 and post-1.18.

why does slimefun needs biome names, you guys arent using anything special for biomes

**Comment by WalshyDev** [comment]
All GEO resources literally rely on biomes. Also, this plugin hosts a very large SDK in which addons may rely on biomes. So is proving this benefits both parties

**Review by WalshyDev — APPROVED** [review]

**Comment by TerslenK** [comment]
> All GEO resources literally rely on biomes. Also, this plugin hosts a very large SDK in which addons may rely on biomes. So is proving this benefits both parties

ohi forgot GEO miner, my bad 😅

**Comment by TheBusyBiscuit** [comment]
Just waiting for a final okay from @Sfiguz7, then we can merge :)

**Review by Sfiguz7 — APPROVED** [review]
LGTM

**Pull request merged**